man/man1/mg_invf_dump.1
.\"------------------------------------------------------------
.\" Id - set Rv,revision, and Dt, Date using rcs-Id tag.
.de Id
.ds Rv \\$3
.ds Dt \\$4
..
.Id $Id: mg_invf_dump.1 16583 2008-07-29 10:20:36Z davidb $
.\"------------------------------------------------------------
.TH mg_invf_dump 1 \*(Dt CITRI
.SH NAME
mg_invf_dump \- Dump out an inverted file in ASCII
.SH SYNOPSIS
.B mg_invf_dump
[
.B \-h
]
[
.B \-b
]
[
.B \-w
]
[
.B \-t
]
[
.BI \-d " directory"
]
.BI \-f " name"
.SH DESCRIPTION
This program dumps out an inverted file produced by the
.BR mg (1)
system as ASCII numbers.  This program could be used in conjunction
with
.BR mgdictlist (1)
to write simple shell programs that work with the inverted files.  The
output from the program looks something like this:
.LP
.RS
.ft 3
.nf
6 337
1
 1 1
1
 5 1
1
 2 1
1
 1 1
4
 1 3
 2 7
 3 2
 5 2
1
 1 1
 . . .
.fi
.ft
.RE
The first number (6) is the number of documents in the collection.  It
is followed by the number of different terms in the collection.  For
each term, the number of different documents it occurs in follows. The
document numbers that the term occurs in are listed next.  If
.B \-w
is specified, then the number of times the term occurred in the
document is printed alongside the document number.
If
.B \-t
is specified, the term itself is also displayed in quotes.
.SH OPTIONS
Options may appear in any order.
.TP "\w'\fB\-d\fP \fIdirectory\fP'u+2n"
.B \-h
This displays a usage line on
.IR stderr .
.TP
.B \-b
This option will cause the output from the program to be in fixed-size
binary numbers, rather than in ASCII.
.TP
.B \-w
If the inverted file is a level-2 or level-3 inverted file, this causes
the word counts per document to be output.
.TP
.B \-t
This option causes each term to be displayed along with the number of
documents it appears in. The term is printed in quotes,
as in "the".
.TP
.BI \-d " directory"
This specifies the directory where the document collection can be found.
.TP
.BI \-f " name"
This specifies the base name of the document collection.
.SH ENVIRONMENT
.TP "\w'\fBMGDATA\fP'u+2n"
.SB MGDATA
If this environment variable exists, then its value is used as the
default directory where the
.BR mg (1)
collection files are.  If this variable does not exist, then the
directory \*(lq\fB.\fP\*(rq is used by default.  The command line
option
.BI \-d " directory"
overrides the directory in
.BR MGDATA .
.SH FILES
.TP 20
.B *.invf.dict
The compressed stemmed dictionary.
.TP
.B *.invf.ORG
The original inverted file saved by
.BR mg_invf_rebuild .
.TP
.B *.invf
The inverted file.
.SH "SEE ALSO"
.na
.BR mg (1),
.BR mg_compression_dict (1),
.BR mg_fast_comp_dict (1),
.BR mg_get (1),
.BR mg_invf_dict (1),
.BR mg_invf_rebuild (1),
.BR mg_passes (1),
.BR mg_perf_hash_build (1),
.BR mg_text_estimate (1),
.BR mg_weights_build (1),
.BR mgbilevel (1),
.BR mgbuild (1),
.BR mgdictlist (1),
.BR mgfelics (1),
.BR mgquery (1),
.BR mgstat (1),
.BR mgtic (1),
.BR mgticbuild (1),
.BR mgticdump (1),
.BR mgticprune (1),
.BR mgticstat (1).


src/text/mg_invf_dump.c
/**************************************************************************
 *
 * mg_invf_dump.c -- Program to dump uot an inverted fil
 * Copyright (C) 1994  Neil Sharman
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * $Id: mg_invf_dump.c,v 1.3 1994/11/29 00:32:01 tes Exp $
 *
 **************************************************************************/

#include "sysfuncs.h"
#include "messages.h"
#include "timing.h"
#include "bitio_m.h"
#include "bitio_m_stdio.h"
#include "bitio_gen.h"

#include "mg_files.h"
#include "locallib.h"
#include "words.h"
#include "invf.h"

extern unsigned long S_btg;

/*
   $Log: mg_invf_dump.c,v $
   * Revision 1.3  1994/11/29  00:32:01  tes
   * Committing the new merged files and changes.
   *
   * Revision 1.2  1994/09/20  04:41:50  tes
   * For version 1.1
   *
 */

static char *RCSID = "$Id: mg_invf_dump.c,v 1.3 1994/11/29 00:32:01 tes Exp $";



static void process_files (char *filename);

int binary = 0;
int word_counts = 0;
int term_dump = 0;


int 
main (int argc, char **argv)
{
  ProgTime start;
  char *dir_name, *file_name = "";
  int ch;
  msg_prefix = argv[0];
  dir_name = getenv ("MGDATA");
  opterr = 0;
  msg_prefix = argv[0];
  while ((ch = getopt (argc, argv, "hbwtf:d:")) != -1)
    switch (ch)
      {
      case 'f':		/* input file */
	file_name = optarg;
	break;
      case 'd':
        set_basepath(optarg);
	break;
      case 'b':
	binary = 1;
	break;
      case 'w':
	word_counts = 1;
	break;
      case 't':
	term_dump = 1;
	break;
      case 'h':
      case '?':
	fprintf (stderr, "usage: %s [-h] [-b] [-w] [-t] [-f input_file]"
		 "[-d data directory]\n", argv[0]);
	exit (1);
      }
  GetTime (&start);
  process_files (file_name);
  Message ("%s\n", ElapsedTime (&start, NULL));
  exit (0);
}

static void 
process_files (char *name)
{
  unsigned long N, Nstatic, k;
  FILE *dict;
  FILE *invf;
  struct invf_dict_header idh;
  struct invf_file_header ifh;


  dict = open_file (name, INVF_DICT_SUFFIX, "r", MAGIC_STEM_BUILD, MG_ABORT);

  fread ((char *) &idh, sizeof (idh), 1, dict);

  if (!(invf = open_file (name, INVF_SUFFIX ".ORG", "r", MAGIC_INVF, MG_CONTINUE)))
    invf = open_file (name, INVF_SUFFIX, "r", MAGIC_INVF, MG_ABORT);

  fread ((char *) &ifh, sizeof (ifh), 1, invf);
  if (ifh.skip_mode != 0)
    FatalError (1, "The invf file contains skips. Unable to dump.");

  DECODE_START (invf)
    N = idh.num_of_docs;
  Nstatic = idh.static_num_of_docs;
  if (binary)
    {
      fwrite ((char *) &N, sizeof (N), 1, stdout);
      fwrite ((char *) &ifh.no_of_words, sizeof (ifh.no_of_words), 1, stdout);
    }
  else
    printf ("%ld %ld\n", N, ifh.no_of_words);
  for (k = 0; k < ifh.no_of_words; k++)
    {
      int i, blk, doc;
      register unsigned long suff, prefix;
      unsigned long fcnt, wcnt;
      char term[MAXSTEMLEN + 1];

      prefix = fgetc (dict);
      suff = fgetc (dict);
      fread (&term[prefix], sizeof (char), suff, dict);
      fread ((char *) &fcnt, sizeof (fcnt), 1, dict);
      fread ((char *) &wcnt, sizeof (wcnt), 1, dict);
      term[prefix + suff] = '\0';

      if (binary)
	fwrite ((char *) &fcnt, sizeof (fcnt), 1, stdout);
      else
	{
	  if (term_dump)
	    printf ("%ld    \"%s\"\n", fcnt, term);
	  else
	    printf ("%ld\n", fcnt);
	}
      blk = BIO_Bblock_Init (Nstatic, fcnt);
      for (doc = i = 0; i < fcnt; i++)
	{
	  int num;
	  BBLOCK_DECODE (num, blk);
	  doc += num;
	  if (binary)
	    fwrite ((char *) &doc, sizeof (doc), 1, stdout);
	  else
	    printf (" %d", doc);
	  if (ifh.InvfLevel >= 2)
	    {
	      int count;
	      GAMMA_DECODE (count);
	      if (word_counts)
		if (binary)
		  fwrite ((char *) &count, sizeof (count), 1, stdout);
		else
		  printf (" %d", count);
	    }
	  if (!binary)
	    putchar ('\n');
	}
      while (__btg)
	DECODE_BIT;
    }
  DECODE_DONE
}
